File: leetcode/0139-word-break/java/solution_3.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        return wordBreak(s, wordDict, new HashMap<>());
    }

    private boolean wordBreak(String s, List<String> wordDict, Map<String, Boolean> memo) {
        
        if (s.isEmpty()) {
            return true;
        }

        if (memo.containsKey(s)) {
            return memo.get(s);
        }

        boolean res = false;
        for (String word : wordDict) {
            if (s.startsWith(word)) {
                res |= wordBreak(s.substring(word.length()), wordDict, memo);
            }
        }

        memo.put(s, res);
        return res;
    }
}

File: leetcode/0139-word-break/rust/solution_3.rs
use std::collections::HashMap;

impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let mut memo: HashMap<u8, bool> = HashMap::new();
        Solution::word_break_search(&s, 0, &word_dict, &mut memo)
    }

    fn word_break_search(s: &String, index: u8, word_dict: &Vec<String>, memo: &mut HashMap<u8,bool>) -> bool {
        if index == s.len() as u8 {
            return true;
        }

        if memo.contains_key(&index) {
            return memo.get(&index).unwrap().clone();
        }

        let mut res = false;
        for word in word_dict {
            if Solution::starts_with_at(s.as_str(), &word, index as usize) {
                res |= Solution::word_break_search(s, index + word.len() as u8, word_dict, memo);
            }
        }

        memo.insert(index, res);
        return res;
    }

    fn starts_with_at(s: &str, prefix: &str, offset: usize) -> bool {
        if offset > s.len() {
            return false;
        }
        s[offset..].starts_with(prefix)
    }
}